activerecord/lib/active_record/reflection.rb

```diff
@@ -503,7 +503,13 @@ def foreign_key(infer_from_inverse_of: true)
           # composite foreign keys support
           options[:query_constraints].map { |fk| fk.to_s.freeze }.freeze
         else
-          -(options[:foreign_key]&.to_s || derive_foreign_key(infer_from_inverse_of: infer_from_inverse_of))
+          derived_fk = options[:foreign_key]&.to_s || derive_foreign_key(infer_from_inverse_of: infer_from_inverse_of)
+
+          if active_record.has_query_constraints?
```

Comment:
We should allow applications to override `primary_key:` configuration and disable automatic foreign key resolution as otherwise a setup like:
```ruby
class Order < ActiveRecord::Base
   self.primary_key = :id
   query_constraints :shop_id, :id
end

class LineItem < ActiveRecord::Base
  belongs_to :order, primary_key: :id, foreign_key: :order_id
  # and
  #  belongs_to :order, primary_key: :id
end
```
will be broken since we will always try to append/prepend `shop_id` to the foreign key

```suggestion
          if active_record.has_query_constraints? && !options[:primary_key]
```

Reply:
I can't make this change to add the primary key check without breaking the existing tests. I originally had it so if you set an explicit foreign key we don't generate a query constraint but then the async destroy tests were failing because it couldn't figure out the query constraints because the derived foreign key was wrong. I think setting an explicit foreign key should always override query constraints - I'll make this change.

Reply:
>I think setting an explicit foreign key should always override query constraints

That works 👍